### ISSUE TITLE
AOT compilation lags process to death

### ISSUE THREAD

**Issue by ninja-** [eclipse/openj9#11157]
I am having an issue where a process running with
- shared classes
- xtune:virtualized
- aot enabled

if there aren't cached classes yet, the app will start normally and then as it becomes idle few seconds later, lag the server to death with AOT compilation requests (most likely), then about 2-3 minutes later it's back to normal and the app is responsive again.
The behaviour seems the same with both jitserver client enabled and disabled...

Is this expected behaviour? And is compiler thread at least set to be lower priority than the app threads?

**Comment by fjeremic** [comment]
@dsouzai FYI.

**Comment by ninja-** [comment]
I will try to check what happens with async-profiler in a couple of days. Needs some work to integrate it with production.

**Comment by dsouzai** [comment]
When running with `-Xtune:virtualized` and AOT, the first run is considered a throwaway. The reason is because `-Xtune:virtualized` enables `forceAOT` which causes almost every<sup>1</sup> compile to be an AOT compile. Additionally, when AOT is enabled we compile methods sooner than we would've if it wasn't. I suppose the combination of these two facts results in a big burst of compilations.

That said, 2-3 minutes does seem excessive. Some questions:

* Do you see the same issue if you ran with `-Xshareclasses:none` and without `-Xtune:virtualized`?
* How many CPUs is the process bound to? 
* Could you can run with `-Xjit:verbose={compilePerformance},vlog=<path to vlog>/vlog` on the first run and upload the verbose log?
* Do you see a lag during the second run?

> And is compiler thread at least set to be lower priority than the app threads?

The comp threads seem to be created with the lowest priority. 

@mpirvu do you have any other insights?

<hr/>
1. An AOT compile can only happen if the class exists in the SCC.

**Comment by ninja-** [comment]
is it possible to get rid of -Xtune:virtualized while keeping all other benefits listed here using other flags?

```
This option is recommended for CPU-constrained environments, such as those found in cloud deployments that use containers. Internally, the option makes the JIT compiler more conservative with inlining and recompilation decisions, which saves CPU resources. The Garbage Collector also reduces the rate of heap expansion, which reduces the memory footprint. These changes to reduce the amount of CPU that is consumed are at the expense of a small loss in throughput.
```

**Comment by mpirvu** [comment]
> The comp threads seem to be created with the lowest priority.

Actually, on Linux compilation threads have the same priority as the application threads. The reason is that the application needs to be run as root to be able to change thread priorities. This is why, while OpenJ9 has an option to change compilation thread priority, it is a no-op on Linux. It does work on Windows though.

@ninja- You described a "cold start" scenario where the application is actually populating the shared class cache (SCC) with classes and AOT bodies.  Subsequent JVM instances attaching to the same SCC should incur much less overhead. I do expect JITServer to give you relief from the compilation overhead though. I would very interested to see the content of a verbose log obtained with the options @dsouzai mentioned above.
Also, are you using the JVMs in containers? If so, you must put the SCC in a volume so that it can be shared by several JVMs. Otherwise, the populated SCC goes away when the container is destroyed.

**Comment by ninja-** [comment]
> @ninja- You described a "cold start" scenario where the application is actually populating the shared class cache (SCC) with classes and AOT bodies. Subsequent JVM instances attaching to the same SCC should incur much less overhead. I do expect JITServer to give you relief from the compilation overhead though. I would very interested to see the content of a verbose log obtained with the options @dsouzai mentioned above.

That's probably the case, but I am doing this test particulat test with one instance only/

> Also, are you using the JVMs in containers? If so, you must put the SCC in a volume so that it can be shared by several JVMs. Otherwise, the populated SCC goes away when the container is destroyed.

No worries that's taken care of :)

**Comment by ninja-** [comment]
@mpirvu putting AOT aside, I had a weird issue lately, and I am pretty sure I haven't change any configuration options in last few days. Servers running with tune:virtualized started having an issue where they wouldn't grow their heaps from initial to maximum even though GC was clearly lagging them to death. Could this be caused by SCC file growing to big size in a few days of use?
Java RSS size is far from container memory limit. It may be close to k8s request, if that's even a thing and is used for some cgroups configuration option (because for example cpu request _is_ used to configure cgroups).

I will try to reproduce this again by wiping production SCC files, in the meantime I changed config to use Xms = Xmx. Weird...

**Comment by mpirvu** [comment]
The JVM does not look at k8s request value, so that cannot be a factor.
The GC expansion is driven by excessive time spent in GC. I don't see a connection to the SCC.
If you can collect a GC verbose log I'll relay it to one of our GC experts. The relevant options are:
`-verbose:gc -Xverbosegclog:path_to_gclog`
See https://www.eclipse.org/openj9/docs/xverbosegclog/ for more configuration of the gc logs.

@amicic

**Comment by ninja-** [comment]
@mpirvu or maybe the fact that SCC is full for JIT and AOT could cause lags?
I'll try to do a proper repro with enabled gc/jit logging but it may take a couple more days...for time being I'll probably try removing tune:virtualized

**Comment by mpirvu** [comment]
The fact that the SCC is full should not have any bad side effects. The compiler just stops adding more classes/AOT methods to the SCC, but the existing ones can still be read. Liberty application server has a 80MB SCC that typically gets full and we haven't observed the negative effects you did. Something else is at play and maybe gclogs could shed some light.

**Comment by ninja-** [comment]
@mpirvu I am almost done implementing automatic profiling on production so no weird JVM cause of lags is going to hide :) should be fun to see the results

**Comment by ninja-** [comment]
@mpirvu when I restart all JVMs at once and use a fresh SCC file, one random JVM instance per machine will be the "victim" and keep AOT compiling everything(I guess there's some cross process mutex that prevents multiple JVMs from compiling methods that will end up in SCC?) while reaching Docker CPU limit (150%) and becoming unresponsive.
So in that case JIT server gets spammed with "AOT warm" requests etc. etc.

Same issue is after restarting JIT server or redirecting traffic to another JIT server, but then it's JIT compilations not AOT.
IMO compiler should detect that it's reaching Docker CPU limit and back off, but I am surprised that just sending and receiving compilations with JIT Server is taking so much CPU...

I wasn't able to get JIT compilePerformance logs yet.

**Comment by mpirvu** [comment]
> I guess there's some cross process mutex that prevents multiple JVMs from compiling methods that will end up in SCC

There is no such thing. Maybe all JVMs compile the exact set of methods in the same order and the first JVM carries all the load. 

> while reaching Docker CPU limit (150%) and becoming unresponsive

I wonder if something else is going on. On linux, compilation threads have the same priority as the application threads. I expect the JVM that performs all the AOT compilations to become very slow, but not unresponsive.

> So in that case JIT server gets spammed with "AOT warm" requests etc. etc

Even if multiple JVMs attempt compilations in parallel, if they are going to be sent to JITServer, the server will see the union of these compilations set. So this scenario will not make it any easier on the JITServer.

> IMO compiler should detect that it's reaching Docker CPU limit and back off

CPU is considered a "compressible" resource, meaning that you can overcommit and things will slow down, but continue to make progress. You can always force the JVMs to use a single compilation thread with `-XcompilationThreads1` and in this case the JVMs cannot be overwhelmed by the compilation activity.

> I am surprised that just sending and receiving compilations with JIT Server is taking so much CPU...

In our experiments, the CPU burnt by the compilation threads at the JVM clients is about 15-20% of the CPU that would be consumed if the JVMs were to perform the compilation themselves. I see this value increasing if all the network traffic goes through proxies.

Again, I would love to see the verbose logs to understand what we are dealing with here.

**Comment by ninja-** [comment]
> There is no such thing. Maybe all JVMs compile the exact set of methods in the same order and the first JVM carries all the load.

So all JVMs are AOT compiling the same methods multiple times, and only when it's time to put it in SCC they throw the AOT code away because it's already in SCC? In such case maybe it could be improved to have a mutex?

**Comment by mpirvu** [comment]
> So all JVMs are AOT compiling the same methods multiple times, and only when it's time to put it in SCC they throw the AOT code away because it's already in SCC? In such case maybe it could be improved to have a mutex?

When a compilation request is generated, the JVM first check to see whether a matching compiled methods already exists in SCC. If it does not, then it checks whether the method is eligible for an AOT compilation. If it is, it will start an AOT compilation and when the compiled body is ready the JVM acquires a lock to "upload" the AOT body to the SCC. There is no information relayed to the other JVMs saying, "I am going to AOT compile this method, please don't try to do the same". Hence, if really unlucky it is possible for other JVMs to AOT compile in parallel for the same method. However, since compilations are taking milliseconds, those JVM would need to be really synchronized for this behavior to happen at scale.
